src/firestore/fsConfig.ts

```diff
@@ -0,0 +1,78 @@
+import { FirebaseError } from "../error";
+import { logger } from "../logger";
+import { Options } from "../options";
+
+export interface ParsedFirestoreConfig {
+  database: string;
+  rules?: string;
+  indexes?: string;
+}
+
+export function getFirestoreConfig(projectId: string, options: Options): ParsedFirestoreConfig[] {
+  const fsConfig = options.config.src.firestore;
+  if (fsConfig === undefined) {
+    return [];
+  }
+
+  const rc = options.rc;
+  let allDatabases = !options.only;
+  const onlyDatabases = new Set<string>();
+  if (options.only) {
+    const split = options.only.split(",");
+    if (split.includes("firestore")) {
+      allDatabases = true;
+    } else {
+      for (const value of split) {
+        if (value.startsWith("firestore:")) {
+          const target = value.split(":")[1];
+          onlyDatabases.add(target);
+        }
+      }
+    }
+  }
+
+  // single DB
+  if (!Array.isArray(fsConfig)) {
+    if (fsConfig) {
+      // databaseId is (default) if none provided
+      const databaseId = fsConfig.database || `(default)`;
+      return [{ rules: fsConfig.rules, indexes: fsConfig.indexes, database: databaseId }];
+    } else {
+      logger.debug("Possibly invalid database config: ", JSON.stringify(fsConfig));
+      return [];
+    }
+  }
+
+  const results: ParsedFirestoreConfig[] = [];
+  for (const c of fsConfig) {
```

Comment:
This block of the function (the loop across the `c` of fsConfig) is a bit confusing IMO. Would it be possible to explain some of the thinking here a bit more?

My apologies.

Reply:
the confusing part here is mostly from logic to support options using `--only`.

target-to-database is a one-to-many mapping that the user can configure using https://firebase.google.com/docs/cli/targets#set-up-deploy-target-hosting

think of it this way, we will only include the config if there is no `--only` clause, or if the database is included in the target mapping passed to the `--only` option
for the target case, we first check that the target mapping exists, and if it is then push the rules/indexes specified for each database in the mapping

the logic is equivalent to https://github.com/firebase/firebase-tools/blob/master/src/database/rulesConfig.ts#L62 used for rtdb (goal here being to be as consistent as possible to how rtdb is configured in firebase.json